File: src/server/HarmonyHub.js
import path from 'path'
import Pyjs from './Pyjs'

export function getId(config, type, name) {
  for (const [key, value] of Object.entries(config[type])) {
    if (type === 'Activities' && value === name) {
      return key
    } else if (type === 'Devices' && key === name) {
      return value.id
    }
  }

  return null
}

export default class HarmonyHub {
  /**
   * @arg {string} hubIP
   * @example
   * const hub = new HarmonyHub('10.0.0...')
   *
   * await hub.connect()
   * //...
   * await hub.disconnect()
   */
  constructor(hubIP) {
    this.pyjs = new Pyjs({
      // debug: true,
      entry: path.join(__dirname, 'HarmonyHub.py'),
      spawnArgs: [hubIP],
    })
  }

  async connect() {
    return this.pyjs.connect()
  }

  async disconnect() {
    return this.pyjs
      .send({ commandName: 'close' })
      .then(() => this.pyjs.disconnect())
  }

  async getConfig() {
    if (!this.DEBUG && this._cachedConfig) {
      return this._cachedConfig
    }

    const result = await this.pyjs.send({ commandName: 'get-config' })

    if (!this.DEBUG) {
      this._cachedConfig = result
    }

    return result
  }

  /**
   * @arg {{
   *   command: string,
   *   delay: number,
   *   device: string,
   *   repeats: number,
   * }} opts
   * @example
   * await hub.sendCommand({
   *   device: 'Samsung TV',
   *   command: 'VolumeUp',
   * })
   */
  async sendCommand(opts) {
    const config = await this.getConfig()
    const deviceId = getId(config, 'Devices', opts.device)

    if (deviceId) {
      opts = {
        delay: 0,
        repeats: 1,
        ...opts,
      }

      opts.deviceId = deviceId
      delete opts.device
      return this.pyjs.send({ commandName: 'send-command', ...opts })
    } else {
      throw new Error(`Device doesn't exist in hub: ${opts.device}`)
    }
  }

  /**
   * @arg {string} activity
   * @example
   * await hub.startActivity('Watch TV')
   */
  async startActivity(activity) {
    const config = await this.getConfig()
    const id = getId(config, 'Activities', activity)

    if (id) {
      return this.pyjs.send({ commandName: 'start-activity', id })
    } else {
      throw new Error(`Activity doesn't exist in hub: ${activity}`)
    }
  }

  /**
   * Stop current activity
   * @example
   * await hub.stopActivity()
   */
  async stopActivity() {
    return this.pyjs.send({ commandName: 'start-activity', id: '-1' })
  }
}

File: src/server/HarmonyHub.py
###
# NOTE: most of this logic was referenced from here:
# https://github.com/ehendrix23/aioharmony/blob/0961024/aioharmony/__main__.py
###

import asyncio
import sys
import aioharmony.exceptions
from aioharmony.harmonyapi import HarmonyAPI, SendCommandDevice
from aioharmony.responsehandler import Handler
from Pyjs.client import on_message

CLIENT = None
HUB_IP = sys.argv[2]

async def getClient(ip_address):
  client = HarmonyAPI(ip_address)

  if await client.connect():
    return client

  return None

# TODO: combine with getClient
async def setGlobalClient():
  global CLIENT

  try:
    CLIENT = await getClient(HUB_IP)
  except aioharmony.exceptions.TimeOut:
    print('Timed out when trying to connect to hub')
    raise

async def handleInput(props):
  commandName = props['commandName']

  if commandName == 'close':
    await CLIENT.close()

  elif commandName == 'get-config':
    return CLIENT.json_config

  elif commandName == 'send-command':
    command = SendCommandDevice(
      device=props['deviceId'],
      command=props['command'],
      delay=props['delay']
    )

    commandList = []
    for _ in range(props['repeats']):
      commandList.append(command)
      if props['delay'] > 0:
        commandList.append(props['delay'])

    resultList = await CLIENT.send_commands(commandList)

    if resultList:
      errorList = []

      for result in resultList:
        errorList.append(result.msg)

      return { 'errors': errorList }

  elif commandName == 'start-activity':
    response = await CLIENT.start_activity(props['id'])

    if not response[0]:
      return { 'errors': [ response[1] ] }

  return {}

if __name__ == '__main__':
  loop = asyncio.new_event_loop()
  asyncio.set_event_loop(loop)
  loop.run_until_complete(setGlobalClient())

  def handler(data, callback):
    result = loop.run_until_complete(asyncio.gather(
      handleInput(data)
    ))
    callback(result[0])

  on_message(handler)
